hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/committer/manifest/stages/RenameFilesStage.java

```diff
@@ -0,0 +1,153 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.mapreduce.lib.output.committer.manifest.stages;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hadoop.mapreduce.lib.output.committer.manifest.files.FileOrDirEntry;
+import org.apache.hadoop.mapreduce.lib.output.committer.manifest.files.ManifestSuccessData;
+import org.apache.hadoop.mapreduce.lib.output.committer.manifest.files.TaskManifest;
+import org.apache.hadoop.util.functional.TaskPool;
+
+import static org.apache.hadoop.mapreduce.lib.output.committer.manifest.ManifestCommitterConstants.SUCCESS_MARKER_FILE_LIMIT;
+import static org.apache.hadoop.mapreduce.lib.output.committer.manifest.ManifestCommitterStatisticNames.OP_STAGE_JOB_COMMIT;
+import static org.apache.hadoop.mapreduce.lib.output.committer.manifest.ManifestCommitterStatisticNames.OP_STAGE_JOB_RENAME_FILES;
+import static org.apache.hadoop.mapreduce.lib.output.committer.manifest.impl.ManifestCommitterSupport.createManifestOutcome;
+import static org.apache.hadoop.thirdparty.com.google.common.collect.Iterables.concat;
+
+/**
+ * This stage renames all the files.
+ * It retuns a manifest success data file summarizing the
+ * output, but does not add iostatistics to it.
+ */
+public class RenameFilesStage extends
+    AbstractJobCommitStage<List<TaskManifest>, ManifestSuccessData> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      RenameFilesStage.class);
+
+  /**
+   * List of all files committed.
+   */
+  private final List<FileOrDirEntry> filesCommitted = new ArrayList<>();
+
+  /**
+   * Total file size.
+   */
+  private long totalFileSize = 0;
+
+  public RenameFilesStage(final StageConfig stageConfig) {
+    super(false, stageConfig, OP_STAGE_JOB_RENAME_FILES, true);
+  }
+
+  /**
+   * Get the list of files committed.
+   * Access is not synchronized.
+   * @return direct access to the list of files.
+   */
+  public synchronized  List<FileOrDirEntry> getFilesCommitted() {
+    return filesCommitted;
+  }
+
+  /**
+   * Get the total file size of the committed task.
+   * @return a number greater than or equal to zero.
+   */
+  public synchronized long getTotalFileSize() {
+    return totalFileSize;
+  }
+
+  /**
+   * Rename files in job commit.
+   * @param taskManifests a list of task manifests containing files.
+   * @return the job report.
+   * @throws IOException failure
+   */
+  @Override
+  protected ManifestSuccessData executeStage(
```

Comment:
As a future enhancement... would it make sense to try rationalizing the files to be moved with an aim to move entire directories instead of individual files?
e.g. ${outputDir}/_temporary/${jobId}/0/taskAttempt01-1/actual-dir/[100s of files]
If the target - presumably always `${outputDir}/actual-dir` does not exist - move ${outputDir}/_temporary/${jobId}/0/taskAttempt01-1/actual-dir/ to ${outputDir}/_temporary/${jobId}/0/taskAttempt01-1/actual-dir instead of moving the files under it individually.

Actually, thinking some more about this - if task directories have the same output dirs created, this is pointless, which is quite likely.
This, then comes to - can taskCommit actually move files to the destination (or in-fact write the files straight to the destination)? I think the answer to this is also a no. There's a risk of the AM/driver not hearing back from a commitTask, after a task commits, and triggering another attempt of the same task.

This, however, is almost equivalent to writing out the manifest file as part of the commit. We end up with a single manifest per task-attempt because the commit renames it to be based on the taskId instead of the taskAttemptId, and this manifest contains information about the actual files to be moved (can think of at least one scenario where this potentially as well - fairly special set of circumstances).

Don't think this facilitates tasks moving files to a common (not task specific) location, or does it (files generated in task attempt locations will likely have the same name ... so if 2 tasks somehow end up being told to commit, they could end up clobbering each others files ... moving the manifest, however, is a single op - so that is why the approach works)

Reply:
May want to ignore this. Turned into writing down though-process. Left if because if helps me refer back to this.